File: Problems/No.2558_Take_Gifts_From_the_Richest_Pile/take_gifts_from_the_richest_pile.py
class Solution(object):
    def pickGifts(self, gifts, k):
        """
        :type gifts: List[int]
        :type k: int
        :rtype: int
        """

        # parameters
        GiftPiles = []
        remaining_gifts = 0

        # start computing
        for initial_gift in gifts:
            heapq.heappush(GiftPiles, (-1)*initial_gift)
        for i in range(k):
            gift = int(((-1)*heapq.heappop(GiftPiles))**0.5)
            heapq.heappush(GiftPiles, (-1)*gift)
        while GiftPiles:
            remaining_gifts += (-1)*heapq.heappop(GiftPiles)

        return remaining_gifts

File: Problems/No.2558_Take_Gifts_From_the_Richest_Pile/take_gifts_from_the_richest_pile.cpp
class Solution {
public:
    long long pickGifts(vector<int>& gifts, int k) {
        
        // parameter
        std::priority_queue<int> GiftPiles;
        long long remaining_gifts = 0;

        // start computing
        for (auto &initial_gift: gifts) GiftPiles.push(initial_gift);
        for (int i=0; i<k; i++) {
            
            auto gift = std::sqrt(GiftPiles.top());
            GiftPiles.pop();
            GiftPiles.push(gift);

        }
        while (!GiftPiles.empty()) {

            remaining_gifts += GiftPiles.top();
            GiftPiles.pop();

        }

        return remaining_gifts;

    }
};
